### PR TITLE
feat(cli/cortex): C-108.x — migrate-config bot.yaml → cortex.yaml (MIG-7.2e)

### PR DESCRIPTION
**Pull request by jcfischer**
## Summary

One-shot CLI that converts a grove-v2-shaped `bot.yaml` to a cortex-shaped `cortex.yaml`, satisfying plan §4 MIG-7.2e. The conversion lives in a pure helper (`migrate-config-lib.ts`) wrapped by a hand-rolled CLI (`migrate-config.ts`, matching the existing `cloud.ts` style).

Closes the conversion side of MIG-7.9 — operator still has to run `cortex start --dry-run` on the produced `cortex.yaml` before flipping the launchd plist.

## Conversion rules (per RESUME §MIG-7.2e)

| Legacy `bot.yaml` | New `cortex.yaml` | Notes |
|---|---|---|
| `agent.operator*` fields | `operator.{id,displayName,discordId,mattermostId,dataResidency}` | operatorId defaults to `agent.name`; dataResidency upcases / falls back to `NZ` |
| `agent.name` + `discord[]` + `mattermost[]` | `agents[]` | Multi-instance synthesizes `${name}`, `${name}-2`, … with a warning |
| `discord[].instanceId` | _dropped_ | Cortex re-derives at runtime per MIG-7.2c-discord-flip; mapping table in `--check` records each legacy id → new presence |
| `personaFile` | `agents[].persona` | Defaults to `./personas/<id>.md` with warning; file existence checked when `configDir` provided |
| `trustedAgentBots[].name` | `agents[].trust[]` | Each name normalized through cortex agent-id regex |
| `api.enabled: true` | `renderers: [{ kind: dashboard, port }]` | Cloud-mode `api.*` fields (endpoint, apiKey, …) drop with warning — re-add via per-network file layout |
| `grove:` block | _dropped_ | F-11 toggles re-implement via renderers later |
| `claude/attachments/execution/github/paths/networksDir/networks/nats/renderers[]` | _pass-through_ | unchanged |

## Validation surface

- Output is parsed through `CortexConfigSchema` inside `convertBotYaml` → the CLI never emits a yaml that cortex itself rejects at load.
- Structural input failures (missing `agent.name`, non-object input) throw eagerly with a pointer at the offending field.
- `--strict` upgrades any warning to exit code 2; otherwise warnings hit stderr and the run exits 0.

## CLI shape

```
bun src/cli/cortex/commands/migrate-config.ts <input.yaml> [--out <out.yaml>] [--check] [--strict]
```

`--check` prints the operator block, agent list, renderer list, instance-mapping table, and warnings without emitting the yaml. `--out FILE` writes to disk (else stdout).

## Test plan

- [x] `bunx tsc --noEmit` clean (worktree)
- [x] `bun test src/cli/cortex/commands/__tests__/migrate-config.test.ts` — 40/40 pass
- [x] Full suite: 2048 pass, 1 fail (pre-existing mission-control React shell test from MIG-2 deferred — known caveat, not from this branch)
- [x] CLI smoke against `full.bot.yaml` fixture — `--check` output matches expectations (1 agent with discord+mattermost presence, dashboard renderer, 2 warnings for `api` + `grove`).
- [x] Fixtures cover the 5 plan-§4 scenarios + argument parsing + operator-id normalization + structural failures + end-to-end exit codes.

## Migration phase

`docs/plan-cortex-migration.md` §4 step 7.2e (line 504). Plan checkbox ticks in a docs PR after this lands (per the convention at commit 7e95a69).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/cli/cortex/commands/migrate-config-lib.ts [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+    contextDepth: inst.contextDepth ?? 10,
+    enableAgentLog: inst.enableAgentLog ?? false,
+    roles: (inst.roles as DiscordPresence["roles"]) ?? [],
+    defaultRole: inst.defaultRole ?? "allow-all",
+    dm: (inst.dm as DiscordPresence["dm"]) ?? {
+      operatorRole: { features: ["chat", "async", "team"], disallowedTools: [], bashGuard: true },
+      defaultRole: "denied" as const,
+      userRoles: [],
+    },
+  } as DiscordPresence;
```
Comment (jcfischer) [inline src/cli/cortex/commands/migrate-config-lib.ts]: **[warning/quality]** Five `as` casts in this file suppress type-checking: `as DiscordPresence` (L247), `as MattermostPresence` (L273), `as DiscordPresence["roles"]` (L240), `as DiscordPresence["dm"]` (L242), `as CortexConfig["renderers"][number]` (L415).

The final `CortexConfigSchema.parse()` at L492 catches runtime mismatches — so these are not exploitable — but they mean the compiler cannot catch a field drift between the manually-built object and the Zod schema. If a required field is added to `DiscordPresenceSchema` tomorrow, TypeScript will stay silent here until the Zod parse throws at runtime.

**Fix option:** Build a partial object and let the schema parse it:
```ts
const partial = { token: inst.token, guildId: String(inst.guildId), ... };
// Let Zod fill defaults + validate:
return DiscordPresenceSchema.parse(partial);
```
This eliminates the manual default duplication (finding #2 below) at the same time.

File: src/cli/cortex/commands/migrate-config-lib.ts [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+ * `${agent.id}-discord` (per RESUME §Decisions §1).
+ */
+function convertDiscordPresence(inst: LegacyDiscordInstance): DiscordPresence {
+  const presence = {
+    enabled: inst.enabled ?? true,
+    token: inst.token,
+    guildId: String(inst.guildId),
+    agentChannelId: String(inst.agentChannelId),
+    logChannelId: String(inst.logChannelId),
+    contextDepth: inst.contextDepth ?? 10,
```
Comment (jcfischer) [inline src/cli/cortex/commands/migrate-config-lib.ts]: **[warning/duplication]** Default values here are manually duplicated from `DiscordPresenceSchema` defaults: `contextDepth ?? 10`, `enableAgentLog ?? false`, `defaultRole ?? "allow-all"`, the entire `dm:` structure (L242-246).

If the Zod schema defaults evolve (e.g., `contextDepth` defaults to 20 in a future PR), the converter will still emit 10 — and because it sets the value *explicitly*, the final `CortexConfigSchema.parse()` will not override it with the new default. The two sources of truth diverge silently.

This is a one-shot migration tool so the blast radius is contained, but it is still latent incorrectness. The fix in finding #1 (parse through `DiscordPresenceSchema.parse(partial)`) eliminates this class of bug entirely — Zod applies its own defaults, single source of truth.

File: src/cli/cortex/commands/migrate-config-lib.ts [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+ * the (possibly multi-entry) `discord[]` and `mattermost[]` arrays.
+ *
+ * Strategy (per RESUME §MIG-7.2e Agents synthesis):
+ *   - Pair discord[i] with mattermost[i] under one agent.
+ *   - Index 0 → base agent id (deriveAgentId(agent.name)).
+ *   - Index ≥1 → `${base}-${i + 1}` (so `luna-2`, `luna-3`).
+ *   - Trust list propagates to every variant.
+ *   - Persona path propagates to every variant.
+ */
+function buildAgents(
```
Comment (jcfischer) [inline src/cli/cortex/commands/migrate-config-lib.ts:359]: **[suggestion/architecture]** `buildAgents` is ~80 lines with three responsibilities: trust-list derivation (L338-352), persona resolution (L354), and the variant construction loop (L357-400). The first two could be extracted as `buildTrustList(legacy, warnings)` and inline without hurting readability, bringing each function under the 25-line suggestion threshold from the architecture lens.

Not a blocker — the linear flow is clear and each section is well-commented. Just a readability note for the next person touching this.

File: src/cli/cortex/commands/migrate-config-lib.ts [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+        legacyKind: "mattermost",
+        legacyIndex: i,
+        legacyInstanceId: m.instanceId,
+        newAgentId: variantId,
+        newPresence: "mattermost",
+      });
+    }
+    agents.push({
+      id: variantId,
+      displayName: variantCount === 1 ? displayName : `${displayName} (${i + 1})`,
```
Comment (jcfischer) [inline src/cli/cortex/commands/migrate-config-lib.ts]: **[nit/quality]** When `variantCount > 1`, the *first* agent's displayName becomes `"Luna (1)"` rather than `"Luna"`. This means the operator's display name *changes* depending on whether they later add a second Discord guild — subtle and potentially surprising. Consider keeping the first variant as the bare `displayName` and only suffixing from index 2 onward:

```ts
displayName: i === 0 ? displayName : `${displayName} (${i + 1})`,
```

Minor — the generated cortex.yaml is a starting point the operator edits anyway.

File: src/cli/cortex/commands/__tests__/migrate-config.test.ts [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+    writeFileSync(bad, "agent: {name: luna, displayName: [unterminated", "utf-8");
+    const code = await runMigrateConfig([bad]);
+    expect(code).toBe(1);
+  });
+
+  test("returns exit code 0 with --help", async () => {
+    const code = await runMigrateConfig(["--help"]);
+    expect(code).toBe(0);
+  });
+});
```
Comment (jcfischer) [inline src/cli/cortex/commands/__tests__/migrate-config.test.ts:598]: **[suggestion/quality]** Good edge case coverage overall, but two gaps:

1. **Empty discord + populated mattermost:** `loadFixture` doesn't have a fixture with `discord: []` and a mattermost entry. The `Math.max(discordInstances.length, mattermostInstances.length, 1)` logic handles it, but a test would lock the behavior — especially the fact that the resulting agent has no discord presence but does have mattermost presence.

2. **Agent name that normalizes to empty string:** e.g., `agent.name: "!!!"` → `deriveAgentId` produces `""` → `AGENT_ID_PATTERN.test("")` returns false → throws. The code handles it correctly but a test would document the contract.

**Comment by jcfischer** [comment]
## Full Review Summary: the-metafactory/cortex#51

**PR:** feat(cli/cortex): C-108.x — migrate-config bot.yaml → cortex.yaml (MIG-7.2e)
**Scope:** 10 files, +1327 / -0 (all additive — lib, CLI, tests, fixtures)

### Lens Results

| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 2 | 2 | 1 | advisory |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | — | — | — | — | n/a (offline CLI) |
| Architecture | 0 | 0 | 1 | 0 | pass |
| EcosystemCompliance | 0 | 0 | 0 | 0 | pass |
| Performance | 0 | 0 | 0 | 0 | pass |
| Duplication | 0 | 1 | 0 | 0 | advisory |

### What I checked and what I found clean

**Security:** No secrets in fixtures (all synthetic tokens). No command injection surface — CLI is pure file I/O. YAML.parse is safe against prototype pollution. `resolvePersona` does `existsSync` only, no file reads from untrusted paths. Trust boundary is the operator's own `bot.yaml` — appropriate for a local migration tool.

**Hardening:** Not applicable — this is a one-shot offline CLI, not a server. No auth layer, no CORS, no rate limiting, no webhooks.

**Ecosystem:** Conventional commit with scope + migration phase ref. Labels correct (`feature`, `now`, `migration`). Co-authored-by present. PR description is excellent — conversion table, CLI shape, test plan with checkmarks. CLAUDE.md doesn't need updating (existing `cli/` bucket covers the new files).

**Performance:** Sync I/O is fine for a one-shot CLI. All loops bounded by input size. No algorithmic concerns.

### Findings (5 total)

1. **[warning/quality]** Five `as` type assertions bypass compiler checking on presence objects (`migrate-config-lib.ts` L247, L273, L240, L242, L415). The final `CortexConfigSchema.parse()` catches runtime mismatches, but the compiler can't flag field drift during development. Fix: parse through `DiscordPresenceSchema.parse(partial)` instead of manual construction + cast.

2. **[warning/duplication]** Default values in `convertDiscordPresence` (contextDepth=10, enableAgentLog=false, defaultRole="allow-all", dm structure) are manually duplicated from `DiscordPresenceSchema` defaults. If schema defaults evolve, the converter silently diverges. Same fix as #1 eliminates this.

3. **[suggestion/architecture]** `buildAgents` is ~80 lines with three responsibilities (trust derivation, persona resolution, variant loop). Could decompose for readability but the linear flow is clear.

4. **[suggestion/quality]** Two untested edge cases: empty discord + populated mattermost, and agent.name that normalizes to empty string. Both are handled correctly by the code — tests would document the contract.

5. **[nit/quality]** Multi-instance displayName: first agent becomes "Luna (1)" instead of "Luna" when `variantCount > 1`. The operator's display name changes depending on instance count — potentially surprising.

### What's solid

- Clean lib/CLI separation — pure conversion logic fully testable without file I/O
- 40 tests across 11 describe blocks with fixture-driven approach
- Output validated through `CortexConfigSchema.parse()` — the CLI can never emit invalid YAML
- Conversion rules match the RESUME §MIG-7.2e spec faithfully
- Warning system with `--strict` mode is a good operator UX pattern
- `--check` dry-run with mapping table gives operators confidence before flipping

### Recommendation

Findings #1 and #2 are related — both go away if the presence converters delegate to `DiscordPresenceSchema.parse()` / `MattermostPresenceSchema.parse()` instead of manual object construction. That's a focused change (~20 lines touched) that eliminates an entire class of drift bugs. The remaining findings are genuine suggestions/nits — address or acknowledge, either way is fine.

**Review by jcfischer — COMMENTED** [review]
Full review (6 lenses + duplication) — well-structured migration CLI with clean lib/CLI separation, excellent test coverage, and faithful spec adherence. Two related warnings around type assertions and duplicated schema defaults; both fixable by delegating presence construction to Zod schema parse.

verdict: blockers=0 majors=2 nits=3 — recommend: request-changes (posted as comment-review — bot account opened the PR; --request-changes blocked by GitHub)

**Comment by jcfischer** [comment]
## Full Review: the-metafactory/cortex#51

### Lens Results
| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 2 | 0 | 0 | advisory |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | 0 | 0 | 0 | 0 | pass (N/A — CLI tool) |
| Architecture | 0 | 0 | 1 | 0 | pass |
| EcosystemCompliance | 0 | 0 | 1 | 0 | pass |
| Performance | 0 | 0 | 0 | 0 | pass |
| Duplication | 0 | 0 | 0 | 1 | pass |

### Findings

**W1 [quality] — Type assertions bypass compile-time validation in presence converters**
`migrate-config-lib.ts` lines 240-247, 271-273, 415: five `as` casts (`as DiscordPresence`, `as MattermostPresence`, `as CortexConfig["renderers"][number]`) suppress type-checking on the conversion output. The final `CortexConfigSchema.parse()` catches invalid data at runtime — so this is not a correctness bug — but when an operator's legacy `roles:` or `dm:` has an unexpected shape, the error trace will point at the top-level schema parse rather than the conversion step that produced the bad data. Consider parsing each presence through its individual schema (`DiscordPresenceSchema.parse(presence)`) at construction time for targeted error messages during migration.

**W2 [quality] — Missing test: mismatched discord/mattermost array lengths**
The `Math.max(discordInstances.length, mattermostInstances.length, 1)` logic at line 357 synthesizes agents where some variants get only one platform presence. Example: 2 discord + 1 mattermost → agent[0] has both, agent[1] has only discord. This is a real deployment shape (adding a second discord guild without a second mattermost instance), and there's no test pinning it. The code looks correct from reading, but the path deserves a fixture.

**S1 [architecture] — `buildAgents` at ~75 lines carries multiple responsibilities**
Trust normalization, persona resolution, variant synthesis, and mapping tracking all live in one function. Not blocking since the function is cohesive and the JSDoc is thorough, but extracting trust normalization and variant construction into named helpers would make the logic easier to audit.

**S2 [ecosystem] — CLAUDE.md not updated with `migrate-config` CLI reference**
The PR adds a new CLI command (`bun src/cli/cortex/commands/migrate-config.ts`). If CLAUDE.md's architecture section lists CLI tooling, this should be mentioned.

**N1 [duplication] — Hand-rolled `parseArgs` duplicates `cloud.ts` pattern**
Both `cloud.ts` and `migrate-config.ts` have hand-rolled argument parsers. Acceptable for two CLIs with different flags, and the PR acknowledges the pattern match. Extract if a third CLI command arrives.

### What I checked and found clean

- **Security**: fixture tokens are synthetic (`discord-token-aaaa`, `mm-token-xxxx`), no secrets in source. CLI reads operator-supplied file paths — expected for a local migration tool, no injection surface.
- **Correctness of conversion rules**: operator extraction, agent-id derivation (`deriveAgentId`), dataResidency normalization, trust normalization, persona resolution, api→dashboard renderer synthesis, grove-block drop warning, pass-through of claude/attachments/execution/github/paths/nats blocks — all match the PR description table and the RESUME §MIG-7.2e spec.
- **Schema round-trip**: output is validated through `CortexConfigSchema.parse()`, so the converter can never emit a YAML that cortex rejects at load. The anti-fields (`agent:`, `trustedAgentBots:`, `discord:`, `mattermost:`) in CortexConfigSchema will catch half-migrated configs.
- **DM defaults alignment**: the hardcoded DM fallback in `convertDiscordPresence` (`operatorRole: {features: ["chat","async","team"], disallowedTools: [], bashGuard: true}, defaultRole: "denied", userRoles: []`) matches `DMConfigSchema.parse({})` exactly.
- **Edge cases**: empty agent name → throws; all-hyphens agent name → throws; mixed-case operatorId → normalizes with warning; malformed dataResidency → falls back to NZ with warning; missing personaFile → defaults to `./personas/<id>.md` with warning.
- **Performance**: `existsSync` is synchronous I/O but acceptable for a one-shot CLI tool — not a hot path.
- **Pattern consistency**: lib/CLI split matches `cloud.ts` established pattern. Pure conversion logic separated from IO. Dependencies flow inward to `common/types/cortex-config`.

File: src/cli/cortex/commands/migrate-config-lib.ts (hunk 1 above)
Comment (jcfischer) [inline src/cli/cortex/commands/migrate-config-lib.ts]: **[warning/quality]** Five `as` casts here and in `convertMattermostPresence` (line 273) bypass compile-time type checking. The output IS validated by `CortexConfigSchema.parse()` downstream, so this is not a correctness bug — but when an operator's legacy `roles:` or `dm:` block has an unexpected shape, the Zod error will point at the top-level schema parse (line 492) rather than this conversion step. Consider `DiscordPresenceSchema.parse(presence)` here instead of `as DiscordPresence` — it gives a targeted error at the point of conversion.

File: src/cli/cortex/commands/migrate-config-lib.ts [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+        message: `trusted bot name "${t.name}" normalized to "${id}"`,
+      });
+    }
+    return id;
+  });
+
+  const persona = resolvePersona(legacy.agent.personaFile, baseId, configDir, warnings);
+  const displayName = legacy.agent.displayName || legacy.agent.name;
+
+  const variantCount = Math.max(discordInstances.length, mattermostInstances.length, 1);
```
Comment (jcfischer) [inline src/cli/cortex/commands/migrate-config-lib.ts:380]: **[warning/quality]** No test for mismatched discord/mattermost array lengths. The `Math.max(discordInstances.length, mattermostInstances.length, 1)` logic here synthesizes agents where some variants get only one platform presence (e.g. 2 discord + 1 mattermost → agent[0] has both, agent[1] has only discord). This is a real deployment shape worth pinning with a fixture.

**Review by jcfischer — COMMENTED** [review]
Full review (6 lenses + duplication) — solid one-shot migration CLI with thorough test coverage and correct schema validation. Two warnings: `as` casts in presence converters weaken error locality, and the mismatched discord/mattermost array length path is untested.

verdict: blockers=0 majors=2 nits=4 — recommend: request-changes (posted as comment-review — bot account opened the PR; --request-changes blocked by GitHub)

**Comment by jcfischer** [comment]
## Full Review Cycle 2: the-metafactory/cortex#51

### Round 1 Findings — Verification

| Finding | Severity | Status |
|---------|----------|--------|
| M1: `as` casts suppress type-checking in presence converters | major | **Fixed** — `convertDiscordPresence` and `convertMattermostPresence` now build a `Record<string, unknown>` candidate and delegate to `DiscordPresenceSchema.parse()` / `MattermostPresenceSchema.parse()`. Zod fills defaults, validates fields, and returns a properly typed object. Both the duplication and the cast escape are eliminated in a single change. Clean. |
| M2: Missing test for mismatched discord/mattermost array lengths | major | **Fixed** — New `mismatched-lengths.bot.yaml` fixture with 2 discord + 1 mattermost. Five tests pin: agent count = max(discord, mm), first agent has both presences, second agent has discord only, empty-discord+populated-mattermost edge case, and schema round-trip. Comprehensive. |
| N1: Extract `buildTrustList` from `buildAgents` | nit | **Fixed** — Standalone `buildTrustList` function at migrate-config-lib.ts:168. Clean extraction, exercised transitively through `convertBotYaml` tests. |
| N2: First-variant displayName should be bare | nit | **Fixed** — `displayName: i === 0 ? displayName : \`\${displayName} (\${i + 1})\`` at migrate-config-lib.ts:262. Tested in "first-variant display name" describe block. |
| N3: Empty-discord + populated-mattermost test | nit | **Fixed** — Inline test at migrate-config.test.ts:638–655. Verifies a single mattermost-only agent with no discord presence. |
| N4: Degenerate agent.name test | nit | **Fixed** — Two tests at migrate-config.test.ts:683–698: `"!!!"` for agent.name and for trustedAgentBots entry. Both verify the throw message. |

### Lenses Applied (6 + duplication)
1. CodeQuality — pass (0 findings)
2. Security — pass (0 findings)
3. Hardening — N/A (one-shot CLI, no API surface)
4. Architecture — pass (0 findings)
5. EcosystemCompliance — pass (0 findings)
6. Performance — pass (0 findings)
7. Duplication — pass (0 findings)

### Surface Covered

- **Conversion correctness:** Traced happy path (minimal, full), multi-discord suffix synthesis, mismatched array lengths, empty+populated cross-platform, trust list normalization, persona resolution with/without configDir, degenerate names, structural failures.
- **Schema fidelity:** `AGENT_ID_PATTERN` matches the cortex-config.ts schema regex (`/^[a-z0-9-]+$/`). Presence converters delegate to the schema's own parse, so field defaults stay in sync.
- **CLI shape:** Arg parser covers positional, `--out`, `--out=`, `--check`, `--strict`, `--help`, unknown flag rejection, missing value detection. `runMigrateConfig` exit codes tested: 0 (success), 1 (input error), 2 (strict + warnings).
- **Duplication:** `parseArgs` in migrate-config.ts vs cloud.ts — different shapes, not real duplication. Presence converters share a pattern but not fields — correct DRY judgment.
- **Security:** Fixtures use synthetic tokens. No secrets, no network calls, no shell commands with user input.

**Review by jcfischer — COMMENTED** [review]
Full review cycle 2 (6 lenses + duplication) — all round 1 findings addressed cleanly. Zod-parse delegation in presence converters eliminates both the `as` cast and default-duplication classes of bug. Mismatched-length fixture + 5 tests lock the multi-platform synthesis logic. 51/51 tests, clean code, no new findings.

verdict: blockers=0 majors=0 nits=0 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)

File: src/cli/cortex/commands/__tests__/migrate-config.test.ts [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+  test("reports `warnings: none` when no warnings produced", () => {
+    const legacy = loadFixture("minimal.bot.yaml");
+    const result = convertBotYaml(legacy, { configDir: FIXTURE_DIR });
+    const report = formatCheckReport(result);
+    if (result.warnings.length === 0) {
+      expect(report).toMatch(/warnings: none/);
+    }
+  });
+});
+
```
Comment (jcfischer) [inline src/cli/cortex/commands/__tests__/migrate-config.test.ts:540]: **[suggestion/quality]** Conditional assertion — this test passes silently if the fixture ever changes to produce warnings. The `if` guard means the expectation is skipped entirely while the test name still says "when no warnings produced."

```typescript
// Current: conditional, can silently pass
if (result.warnings.length === 0) {
  expect(report).toMatch(/warnings: none/);
}

// Suggested: unconditional, fails loudly on fixture change
expect(result.warnings).toHaveLength(0);
expect(report).toMatch(/warnings: none/);
```

**Comment by jcfischer** [comment]
## Full Review: the-metafactory/cortex#51 — Round 2

Holly here. Ran all 6 lenses + duplication against the Round 2 push. All Round 1 findings (2 majors, 4 nits) have been addressed — the Zod-parse-for-defaults fix is clean, the mismatched-lengths fixture locks the synthesis edge case, `buildTrustList` extraction reads well, bare displayName for variant 0 is correct, and both new edge-case tests (empty-discord/populated-mattermost, degenerate-name) are solid.

### Lens Results
| Lens | Critical | Warning | Suggestion | Nit | Verdict |
|------|----------|---------|------------|-----|---------|
| CodeQuality | 0 | 0 | 1 | 0 | pass |
| Security | 0 | 0 | 0 | 0 | pass |
| Hardening | — | — | — | — | N/A (CLI tool) |
| Architecture | 0 | 0 | 0 | 0 | pass |
| EcosystemCompliance | 0 | 0 | 0 | 1 | pass |
| Performance | — | — | — | — | N/A (CLI tool) |
| Duplication | 0 | 0 | 0 | 0 | pass |

### Findings

1. **[suggestion/quality]** `migrate-config.test.ts:532-540` — conditional assertion in "warnings: none" test can silently pass if fixture changes. See inline comment.

2. **[nit/ecosystem]** `fixup(migrate-config)` commit — `fixup` is not a standard conventional commit type. Pre-merge squash folds it away, so cosmetic only.

### What I checked and found clean

- **Zod-parse path**: Both `convertDiscordPresence` and `convertMattermostPresence` build `Record<string, unknown>` candidates and delegate to `DiscordPresenceSchema.parse()` / `MattermostPresenceSchema.parse()`. No `as` casts, defaults flow from the schema. Traced edge cases: `contextDepth` from YAML (number vs string), `dm: {}` (empty-default path), `enabled: false` (carried forward, not defaulted).
- **Synthesis logic**: `buildAgents` correctly pairs discord[i] with mattermost[i], handles `max(discord.len, mattermost.len, 1)` for variant count, propagates trust and persona to all variants. Zero-presence input correctly fails at `AgentSchema.refine` (presence requires ≥1 block).
- **Security**: Fixture tokens are clearly fake. Converter emits field names in warnings, not values, so real secrets stay out of stderr. No injection surfaces (CLI reads YAML, writes YAML).
- **Architecture**: Pure lib + CLI wrapper separation is clean. Dependencies flow inward. No circular imports. `LegacyBotYaml` as a permissive interface (not Zod) is correct — real bot.yaml files carry dropped fields.
- **Duplication**: `toArray`, `coerceString`, `deriveAgentId` are all unique to this module. The `convertDiscord/MattermostPresence` pattern similarity is intentional (different field sets → different functions).
- **51 tests across 14 describe blocks**: happy paths, structural failures, normalization, degenerate input, mismatched lengths, schema round-trips, CLI end-to-end smoke.

**Review by jcfischer — COMMENTED** [review]
Full review (6 lenses + duplication) — Round 2 is clean. All Round 1 findings addressed. One conditional-assertion suggestion and one commit-format nit remain — neither blocks merge.

verdict: blockers=0 majors=0 nits=2 — recommend: merge (posted as comment-review — bot account opened the PR; --approve blocked by GitHub)
